### PR TITLE
Add Force Remove Vesting for Vesting Pallet

### PR DESCRIPTION
**Pull request by NZT48**
There is already a function inside the Vesting Pallet for removing vesting schedules. This PR exposes it through a ROOT call, so that developers and governance can have access to such things.

This PR is related to #11863 issue.

### CODE REVIEW

File: frame/vesting/src/lib.rs [hunk 1]
```diff
@@ -439,6 +439,25 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Force remove a vesting schedule
+		///
+		/// The dispatch origin for this call must be _Root_.
+		///
+		/// - `target`: The account that received that vesting transfer
```
Comment (shawntabrizi) [inline frame/vesting/src/lib.rs:447]:
```suggestion
		/// - `target`: An account that has a vesting schedule.
```

File: frame/vesting/src/tests.rs [hunk 2]
```diff
@@ -1155,3 +1155,39 @@ fn vested_transfer_less_than_existential_deposit_fails() {
...
+		let user3_free_balance = Balances::free_balance(&3);
+		let user4_free_balance = Balances::free_balance(&4);
+		assert_eq!(user3_free_balance, 256 * 30);
+		assert_eq!(user4_free_balance, 256 * 40);
+		// Account 4 should not have any vesting yet.
+		assert_eq!(Vesting::vesting(&4), None);
+		// Make the schedule for the new transfer.
+		let new_vesting_schedule = VestingInfo::new(
+			256 * 5,
+			64, // Vesting over 20 blocks
```
Comment (shawntabrizi) [inline frame/vesting/src/tests.rs:1171]: 20? 64?

File: frame/vesting/src/benchmarking.rs [hunk 3]
```diff
@@ -375,6 +375,49 @@ benchmarks! {
...
+
+		let source: T::AccountId = account("source", 0, SEED);
+		let source_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(source.clone());
+		T::Currency::make_free_balance_be(&source, BalanceOf::<T>::max_value());
+
+		let target: T::AccountId = account("target", 0, SEED);
+		let target_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(target.clone());
+		
+		// Give target existing locks
+		add_locks::<T>(&target, l as u8);
```
Comment (ggwpez) [inline frame/vesting/src/benchmarking.rs:390]: Please add an `assert!(l <= u8::MAX.into());` here, otherwise the benchmarking logic could get messed up.  
The `add_locks` could also be re-written to accept an `u32`, but i think that is out of scope.

File: frame/vesting/src/benchmarking.rs [hunk 4]
```diff
@@ -375,6 +375,49 @@ benchmarks! {
...
+		// It will remove last vesting schedule
+		let schedule_index = 2;
+
+		let vesting_schedule = VestingInfo::new(
+			transfer_amount,
+			per_block,
+			1u32.into(),
+		);
+		// Add max vesting schedules
+		for _ in 0..T::MAX_VESTING_SCHEDULES {
```
Comment (ggwpez) [inline frame/vesting/src/benchmarking.rs:405]:
```suggestion
		for _ in 0..s {
```
Or can you re-use the `add_vesting_schedules` function?

File: frame/vesting/src/benchmarking.rs [hunk 5]
```diff
@@ -375,6 +375,49 @@ benchmarks! {
...
+
+		let target: T::AccountId = account("target", 0, SEED);
+		let target_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(target.clone());
+		
+		// Give target existing locks
+		add_locks::<T>(&target, l as u8);
+		
+		let per_block = T::MinVestedTransfer::get();
+		let transfer_amount = per_block.checked_mul(&20u32.into()).unwrap();
+		// 2 x transfer_amount because we will create 3 vesting schedules and remove one
```
Comment (ggwpez) [inline frame/vesting/src/benchmarking.rs:394]: Why *3* vesting schedules?

File: frame/vesting/src/benchmarking.rs [hunk 6]
```diff
@@ -375,6 +375,49 @@ benchmarks! {
...
+		for _ in 0..T::MAX_VESTING_SCHEDULES {
+			assert_ok!(Vesting::<T>::do_vested_transfer(
+				source_lookup.clone(),
+				target_lookup.clone(),
+				vesting_schedule.into()
+			));
+		}
+	}: _(RawOrigin::Root, target_lookup, schedule_index)
+	verify {
+		assert_eq!(
```
Comment (ggwpez) [inline frame/vesting/src/benchmarking.rs:414]: This assert fails when I try it...

File: frame/vesting/src/tests.rs (hunk 2 above)
Comment (ggwpez) [reply frame/vesting/src/tests.rs:1171]: The ED is set to 256. I think its more understandable if you use `ED` instead of 256 and write `(ED * 5) / 20`.

File: frame/vesting/src/tests.rs [hunk 7]
```diff
@@ -1155,3 +1155,39 @@ fn vested_transfer_less_than_existential_deposit_fails() {
...
+			10,
+		);
+		assert_ok!(Vesting::vested_transfer(Some(3).into(), 4, new_vesting_schedule));
+		// Now account 4 should have vesting.
+		assert_eq!(Vesting::vesting(&4).unwrap(), vec![new_vesting_schedule]);
+		// Account 4 has 5 * 256 locked.
+		assert_eq!(Vesting::vesting_balance(&4), Some(256 * 5));
+		// Verify only root can call
+		assert_noop!(Vesting::force_remove_vesting_schedule(Some(4).into(), 4, 0), BadOrigin);
+		// Verifies that vesting schedule is removed
```
Comment (ggwpez) [inline frame/vesting/src/tests.rs:1181]:
```suggestion
		// Verify that root can remove the schedule.
```

File: frame/vesting/src/tests.rs [hunk 8]
```diff
@@ -1155,3 +1155,39 @@ fn vested_transfer_less_than_existential_deposit_fails() {
...
+		assert_ok!(Vesting::force_remove_vesting_schedule(RawOrigin::Root.into(), 4, 0));
```
Comment (ggwpez) [inline frame/vesting/src/tests.rs:1182]: This should emit an event, or?  
You can get it with `System::events().last().unwrap()` and then assert that its the expected one.

File: frame/vesting/src/benchmarking.rs [hunk 9]
```diff
@@ -375,6 +375,49 @@ benchmarks! {
 		);
 	}
 
+	force_remove_vesting_schedule {
+		let l in 0 .. MaxLocksOf::<T>::get() - 1;
+		let s in 0 .. T::MAX_VESTING_SCHEDULES;
```
Comment (ggwpez) [inline frame/vesting/src/benchmarking.rs:380]: Just an observation:  
We have multiple benchmarks with components that dont need dynamic inputs, like `T::WeightInfo::force_remove_vesting_schedule(MaxLocksOf::<T>::get(), T::MAX_VESTING_SCHEDULES)`.   
The only advantage that comes to my mind from this is that it makes it possible for devs to re-use the Substrate weights without benchmarking themselves...  
Otherwise it just slows down our benchmarking.

File: frame/vesting/src/benchmarking.rs (hunk 3 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs:391]: Added, thanks :)

File: frame/vesting/src/benchmarking.rs (hunk 5 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs:395]: That is the configured number for MAX_VESTING_SCHEDULES for tests, but I have updated this part to include variable that calculates number of expected vestings after removal

File: frame/vesting/src/tests.rs (hunk 2 above)
Comment (NZT48) [reply frame/vesting/src/tests.rs]: @ggwpez done, thanks

File: frame/vesting/src/benchmarking.rs (hunk 4 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs:407]: My inital solution was using `add_vesting_schedules` function, but there was some bug that when I use this function then the verify block and removing function is executed multiple times, I discussed this with @shawntabrizi and he said to use this solution that works, because the bug shouldn't be related to this test. 

If you write `s` instead of `T::MAX_VESTING_SCHEDULES`, it will not execute function inside the for loop and will break the test, I will try to figure out why.

File: frame/vesting/src/benchmarking.rs (hunk 6 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs:416]: Can you please provide me command that you use to run the test, so I can recreate it, it works for me when I run: ```cargo test -p pallet-vesting --features runtime-benchmarks```

File: frame/vesting/src/tests.rs (hunk 7 above)
Comment (NZT48) [reply frame/vesting/src/tests.rs]: Fixed, thanks for feedback

File: frame/vesting/src/tests.rs (hunk 8 above)
Comment (NZT48) [reply frame/vesting/src/tests.rs:1182]: I was looking around other pallet and it looks to me that this is better way of doing this check:
```rust
System::assert_last_event(
	Event::VestingCompleted {
		account: 4
	}
	.into(),
);
```
If you agree, I will implement it this way.

File: frame/vesting/src/benchmarking.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/vesting/src/benchmarking.rs:395]: @NZT48 you have to write benchmarks for a generic runtime. Yes the test environment may only have 3, but then a different runtime would maybe have more or less. You should always retrieve these values from the generic arguments.

File: frame/vesting/src/benchmarking.rs (hunk 4 above)
Comment (ggwpez) [reply frame/vesting/src/benchmarking.rs:407]: Probably since the `s` parameter starts at 0. You can let it start at 1 instead, otherwise there is no vesting schedule.

File: frame/vesting/src/benchmarking.rs (hunk 6 above)
Comment (ggwpez) [reply frame/vesting/src/benchmarking.rs:416]: Yea its fixed now. It failed with this command on commit 68db03af7b7:   
```pre
cargo run --release --features=runtime-benchmarks --bin substrate -- benchmark pallet --dev --steps=50 --repeat=20 --pallet="pallet-vesting" --extrinsic=""
```
The differences are that the normal `cargo test` uses the mock runtime, whereas this command uses the template node runtime.  
Also this command runs it for 50 steps, instead of just one. So the parameters get instantiated with different values.

File: frame/vesting/src/tests.rs (hunk 8 above)
Comment (ggwpez) [reply frame/vesting/src/tests.rs:1182]: Yes thats even better :smile:

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime dev pallet_vesting

**Comment by ggwpez** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_vesting

File: frame/vesting/src/tests.rs (hunk 8 above)
Comment (NZT48) [reply frame/vesting/src/tests.rs:1182]: This has been added in [7479a1](https://github.com/paritytech/substrate/pull/11910/commits/7479a1212ba53cb5c8e17003f59a0d1835a1bbcf) commit.

File: frame/vesting/src/benchmarking.rs (hunk 5 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs]: Update to determine last vesting schedule index from MAX_VESTING_SCHEDULES that is read from runtime.

File: frame/vesting/src/benchmarking.rs (hunk 4 above)
Comment (NZT48) [reply frame/vesting/src/benchmarking.rs]: I have updated `let s in 2 .. T::MAX_VESTING_SCHEDULES;` to make it work with your suggestion above.
